src/main/haskell/kore/src/Kore/Reflect.hs

```diff
@@ -0,0 +1,386 @@
+
+For any instance of Generic, if the data's "children" (see below) are
+Reflectable, the 'reflect' fuction is automatically implemented and the
+derivation looks something like:
+
+@
+instance
+    (constraints-to-make-MyData-generic)
+    => (Reflectable MyData)
+@
+
+A data type's children are, roughly, the parts that are not declared in
+that data's declaration. Examples:
+
+@
+data MyData = MyData T1 T2
+-- Children: T1 and T2
+
+data MyData = MyData {f1 :: T1, f2 :: T2}
+-- Children: T1 and T2
+@
+-}
+class Reflectable a where
+    reflect :: a -> RecursiveData
+    default reflect
+        :: (Generic a, Reflectable' (Generics.Rep a)) => a -> RecursiveData
+    reflect x = reflect' (Generics.from x)
+
+instance {-# OVERLAPS #-} Reflectable String where
+    reflect = Fix . Value . show
+
+instance Reflectable Text where
+    reflect = Fix . Value . show
+
+instance Reflectable Char where
+    reflect = Fix . Value . show
+
+instance Reflectable Int where
+    reflect = Fix . Value . show
+
+instance Reflectable Integer where
+    reflect = Fix . Value . show
+
+instance Reflectable Natural where
+    reflect = Fix . Value . show
+
+instance Reflectable () where
+    reflect = Fix . Value . show
+
+instance Reflectable (Const Void a) where
+    reflect a = case a of {}
+
+instance {-# OVERLAPPABLE #-} (Reflectable a) => Reflectable [a] where
+    reflect = Fix . List . map reflect
+
+instance (Reflectable a, Reflectable b) => Reflectable (Map.Map a b) where
+    reflect = Fix . List . map reflect . Map.toList
+
+instance (Reflectable a) => Reflectable (Set.Set a) where
+    reflect = Fix . List . map reflect . Set.toList
+
+instance (Reflectable a) => Reflectable (Seq a) where
+    reflect = Fix . List . map reflect . Foldable.toList
+
+instance (Reflectable a1, Reflectable a2) => Reflectable (a1, a2) where
+    reflect (a1, a2) = Fix $ Tuple [reflect a1, reflect a2]
+
+instance (Reflectable a1, Reflectable a2, Reflectable a3)
+    => Reflectable (a1, a2, a3)
+  where
+    reflect (a1, a2, a3) = Fix $ Tuple [reflect a1, reflect a2, reflect a3]
+
+instance (Reflectable a1, Reflectable a2, Reflectable a3, Reflectable a4)
+    => Reflectable (a1, a2, a3, a4)
+  where
+    reflect (a1, a2, a3, a4) =
+        Fix $ Tuple [reflect a1, reflect a2, reflect a3, reflect a4]
+
+instance (Reflectable a) => Reflectable (Maybe a) where
+    reflect (Just a) = mkSum "Just" [reflect a]
+    reflect Nothing = mkSum "Nothing" []
+
+instance (Reflectable a1, Reflectable a2) => Reflectable (Either a1 a2) where
+    reflect (Left a) = mkSum "Left" [reflect a]
+    reflect (Right a) = mkSum "Right" [reflect a]
+
+instance
+    (Reflectable (a (Cofree a b)), Reflectable b)
+    => Reflectable (Cofree a b)
+  where
+    reflect (CofreeT (Identity (b :< a))) =
+        mkSum "Cofree" [reflect b, reflect a]
+
+-----------------------------------
+--      Generic reflection       --
+-----------------------------------
+
+class Reflectable' f where
```

Comment:
Could you add a comment that this class should only be instantiated for `GHC.Generics` types?